### ISSUE TITLE
Sik-Ho Tang | Review -- Colorful Image Colorization (Self-Supervised Learning).

### ISSUE THREAD

**Issue by NorbertZheng** [NorbertZheng/read-papers#123]
Sik-Ho Tang. [Review — Colorful Image Colorization (Self-Supervised Learning)](https://sh-tsang.medium.com/review-colorful-image-colorization-self-supervised-learning-c2667640bd0a).

**Comment by NorbertZheng** [comment]
## Overview
Colorization as Pretext Task in Self-Supervised Learning, Outperforms [Context Prediction](https://sh-tsang.medium.com/review-unsupervised-visual-representation-learning-by-context-prediction-self-supervised-51a1d7ce6aff) & [Context Encoders](https://sh-tsang.medium.com/review-context-encoders-feature-learning-by-inpainting-bd181e48997).

![image](https://github.com/NorbertZheng/read-papers/assets/37235023/84b7ae43-4b67-4fc2-bd86-1a5d71b59c6e)
Example Input Grayscale Photos and Output Colorizations.

In this story, Colorful Image Colorization, by University of California, Berkeley, is reviewed. In this paper:
 - A fully automatic approach is designed, which produces vibrant and realistic colorizations for a grayscale image.
 - This proposed colorization task is treated as a powerful pretext task for self-supervised feature learning, acting as a **cross-channel encoder**.

It is a paper in 2016 ECCV with over 1900 citations.

**Comment by NorbertZheng** [comment]
Multi-Modality!!!

**Comment by NorbertZheng** [comment]
## Colorful Image Colorization
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/7baacd6a-1fec-4185-bb3a-1ad2c703bc7d)
Colorful Image Colorization: Network Architecture.
 - **The color space used is CIE Lab color space.**
 - Given an input lightness channel $X$ (L), the objective is to learn a mapping to the two associated color channels $Y$ (ab).

Each conv layer refers to a block of 2 or 3 repeated conv and ReLU layers, followed by a BatchNorm layer. The net has no pool layers, only spatial downsampling or upsampling is used between conv blocks if needed.

### L2 Loss is NOT Robust
Using naïve L2 loss as shown below is **not robust** to the **inherent ambiguity** and **multimodal nature** of the colorization problem:
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/5325f9fe-f155-47f2-a891-aaf1f28b626f)
For example,
 - Multiple possible solutions: If an object can take on a set of distinct ab values, the optimal solution to the Euclidean loss will be the mean of the set. **This averaging effect favors grayish, desaturated results!!!**
 - Also due to multiple possible solutions: Additionally, if the set of plausible colorizations is non-convex, **the solution will in fact be out of the set, giving implausible results**.

**Comment by NorbertZheng** [comment]
Multiple possible solutions lead to
 - the optimal solution converges to the averaging solution, favoring grayish, desaturated results.
 - the optimal solution is implausible, which is out of such set, if non-convex.

**Comment by NorbertZheng** [comment]
### Multinomial Classification Loss
The problem is treated as a multinomial classification.
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/0edea37a-75bc-4f1c-8926-6bb30ef10b22)
Quantized ab color space with a grid size of 10

The ab output space is quantized into bins with grid size 10 and keeps the $Q=313$ values which are in-gamut, as above.

For a given input $X$, we learn a mapping $\hat{Z}=G(X)$ to a probability distribution over possible colors:
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/b22563a7-918f-498d-a402-e9ef48b50d29)
where $Q$ is the number of quantized ab values.

$Z$ is the vector converted from the ground truth color $Y$, using a soft-encoding scheme (i.e. not exactly one-hot, easier to learn meaningful embeddings!!!):
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/b74e5e4e-cf20-4820-958a-e4c331ff15c9)

The 5-nearest neighbors to $Y_{h,w}$ in the output space are selected and weighted proportionally to their distance from the ground truth using a Gaussian kernel with $\sigma=5$.

Thus, the multinomial cross-entropy loss is:
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/1e4746a7-173f-4308-a6a7-d316c4b9f308)
where $v(\cdot)$ is a weighting term that can be used to rebalance the loss based on color-class rarity.
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/caf540d9-5023-49a8-aeb3-61597e6893fd)
Each pixel is weighted by factor $w$ based on its closest ab bin, and $\lambda=1/2$.
 - The reason to reweight is that the distribution of ab values in natural images is strongly biased towards values with low ab values, due to the appearance of backgrounds such as clouds, pavement, dirt, and walls.
 - The ground truth using loss function is dominated by desaturated ab values.

**To be brief, the class-imbalance problem is achieved by reweighting the loss of each pixel at train time based on the pixel color rarity as above.**

Finally, we map probability distribution $\hat{Z}$ to color values $\hat{Y}$ with function:
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/98bbddac-6bff-4735-81ce-b66946abd89e)
This mapping $H$ is mentioned as below in Section 1.3.

**Comment by NorbertZheng** [comment]
Use classification loss to reduce the difficulty of the pretext task???

**Comment by NorbertZheng** [comment]
### Class Probabilities to Point Estimates
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/9d726257-a622-49d3-9737-7dfbce79058b)
Effect of temperature parameter T on the annealed-mean output.

$H$ is defined to map the predicted distribution $\hat{Z}$ to point estimate $\hat{Y}$ in ab space:
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/7dc47ecd-ee9e-49a4-ba93-d0ffa7466f05)
 - $T=0$: One choice is to take the mode of the predicted distribution for each pixel. But this provides a vibrant but sometimes spatially inconsistent result.
 - $T=1$: On the other hand, taking the mean of the predicted distribution produces spatially consistent but desaturated results.

The temperature $T=0.38$, shown in the middle column of the above figure, captures the vibrancy of the mode while maintaining the spatial coherence of the mean.

(**The introduction of temperature $T$ into Softmax** can be referenced from [Model Distillation](https://medium.com/nerd-for-tech/review-model-distillation-distilling-the-knowledge-in-a-neural-network-image-classification-48ce0c81618a))

Hence, the final system $F$ is the composition of CNN $G$, which produces a predicted distribution over all pixels, and the annealed-mean operation $H$, which produces a final prediction. **(The system is NOT quite end-to-end trainable.)**

**Comment by NorbertZheng** [comment]
NOT end-to-end trainable!!!

**Comment by NorbertZheng** [comment]
## Colorization Results
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/924609bd-49dd-4699-aab5-5fed1c449062)
Colorization results on 10k images in the ImageNet validation set.

The 1.3M (i.e. 1'300'000) images from the ImageNet training set are used for training, the first 10k (i.e. 10'000) images in the ImageNet validation set are used for validation, and a separate 10k (i.e. 10'000) images in the validation set are used for testing.

 - Ours (full): Proposed network with classification loss and class rebalancing.
 - Ours (class): Proposed network with classification loss only, no class rebalancing. ($\lambda=1)
 - Ours (L2): Proposed network with L2 regression loss, trained from scratch.
 - **Ours (L2, ft): Proposed network with L2 regression loss, fine-tuned from full classification with rebalancing network.**

### Perceptual realism (AMT)
A real vs. fake two-alternative forced choice experiment is ran on Amazon Mechanical Turk (AMT). 40 Participants were asked to click on the photo they believed contained fake colors.

The proposed full algorithm fooled participants on 32% of trials, which is significantly higher than all compared algorithms. These results validate the effectiveness of using both
 - classification loss,
 - **and class rebalancing**.

**Comment by NorbertZheng** [comment]
Class rebalancing is just like an adversarial loss with an inductive bias towards colorization!!!

**Comment by NorbertZheng** [comment]
### Semantic Interpretability ([VGG](https://medium.com/coinmonks/paper-review-of-vggnet-1st-runner-up-of-ilsvlc-2014-image-classification-d02355543a11) Classification)
It is tested by feeding our fake colorized images to a [VGG](https://medium.com/coinmonks/paper-review-of-vggnet-1st-runner-up-of-ilsvlc-2014-image-classification-d02355543a11). If the classifier performs well, that means the colorizations are accurate.

Classifier performance drops from 68.3% to 52.7% after ablating colors from the input. After re-colorizing using our full method, the performance is improved to 56.0%.

**Comment by NorbertZheng** [comment]
Use machines, instead of humans, to check the generative performance.

**Comment by NorbertZheng** [comment]
### Legacy Black and White Photos
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/0240a5fb-7ded-46b7-8150-837cc9e09d92)
Applying the proposed method to legacy black and white photos.

The proposed model is still able to produce good colorizations, even though **the low-level image statistics of the legacy photographs are quite different from those of the modern-day photos**.

**Comment by NorbertZheng** [comment]
Zero-shot!!!

**Comment by NorbertZheng** [comment]
### More Examples
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/7ee13043-efc6-41a8-9a16-a798d2208d28)
More examples and results are in the appendix of the paper.

**Comment by NorbertZheng** [comment]
## Self-Supervised Learning Results
The colorization approach serves as a pretext task for representation learning.

The network model is akin to an autoencoder, except that the input and output are different image channels, suggesting the term cross-channel encoder.

![image](https://github.com/NorbertZheng/read-papers/assets/37235023/8177ca8e-8fe3-4c21-a18f-b106208ed5d0)
Left: ImageNet Linear Classification, Right: PASCAL Tests.

### ImageNet Classification
The pre-trained networks are frozen and linear classifiers are learnt on top of the each convolutional layer for ImageNet classification.

Pretraining are performed without semantic label information.

AlexNet directly trained on ImageNet classification achieves the highest performance, and serves as the ceiling for this test.

The proposed method outperforms Gaussian, k-means, and [Context Encoders](https://sh-tsang.medium.com/review-context-encoders-feature-learning-by-inpainting-bd181e48997) [10].

 - **Solving the colorization task encourages representations that linearly separate semantic classes in the trained data distribution.**

**Comment by NorbertZheng** [comment]
### PASCAL Classification
The network is trained by freezing the representation up to certain points, and fine-tuning the remainder.

The network is effectively only able to interpret grayscale images.

### PASCAL Detection
[Fast R-CNN](https://medium.com/coinmonks/review-fast-r-cnn-object-detection-a82e172e87ba) framework is used.

The proposed method outperforms k-means. However, it is inferior to [Context Prediction](https://sh-tsang.medium.com/review-unsupervised-visual-representation-learning-by-context-prediction-self-supervised-51a1d7ce6aff) [14].

### PASCAL Segmentation
[FCN](https://towardsdatascience.com/review-fcn-semantic-segmentation-eb8c9b50d2d1) architecture is used.

The proposed grayscale fine-tuned network achieves performance of 35.0%, approximately equal to Donahue et al. [16], and **adding in color information increases performance to 35.6%** (how???).

**Comment by NorbertZheng** [comment]
By learning the colorization as pretext task without ground truth labels, useful features are learnt, which can be used for downstream tasks, such as image classification, detection, and segmentation.

**Comment by NorbertZheng** [comment]
## Reference
 - [2016 ECCV] [Colorization] [Colorful Image Colorization](https://arxiv.org/abs/1603.08511).
